File: frontend/src/types/qcds.ts
/** QCDS関連の型定義。 */

export type QCDSCategory = "subcontract" | "material" | "other";
export type ExpenseSection = "B_site" | "B_dept" | "C";

export interface ExpenseItemInput {
  id?: string;
  section: ExpenseSection;
  row_no: number;
  system_key?: string | null;
  item_name: string;
  formula_description?: string | null;
  amount_override?: number | null;
  is_custom?: boolean;
}

export interface ExpenseItemRead {
  id: string;
  qcds_id: string;
  section: ExpenseSection;
  row_no: number;
  system_key: string | null;
  item_name: string;
  formula_description: string | null;
  amount_override: number | null;
  is_custom: boolean;
}

export const CATEGORY_LABEL: Record<QCDSCategory, string> = {
  subcontract: "外注",
  material: "資材",
  other: "その他",
};

export interface DirectWorkInput {
  row_no: number;
  work_type?: string | null;
  vendor_id?: string | null;
  vendor_name_snapshot?: string | null;
  category?: QCDSCategory | null;
  budget_amount?: number | null;
  agreed_amount?: number | null;
  settlement_amount?: number | null;
  agreement_checked?: boolean;
  payment_month_4?: number | null;
  payment_month_5?: number | null;
  payment_month_6?: number | null;
  payment_month_7?: number | null;
  payment_month_8?: number | null;
  payment_month_9?: number | null;
  payment_month_10?: number | null;
  payment_month_11?: number | null;
  payment_month_12?: number | null;
  payment_month_1?: number | null;
  payment_month_2?: number | null;
  payment_month_3?: number | null;
  payment_completed?: boolean;
  note?: string | null;
}

export interface DirectWorkRead extends DirectWorkInput {
  id: string;
  vendor_name?: string | null;
  source_scan_result_id?: string | null;
}

export interface QCDSInput {
  revision?: number;
  spare_cost?: number | null;
  industrial_waste_cost?: number | null;
  labor_insurance_rate?: number;
  construction_insurance_rate?: number;
  special_insurance_rate?: number;
  office_supplies?: number;
  communication_cost?: number;
  misc_cost?: number;
  site_staff_salary_rate?: number;
  common_overhead_rate?: number | null;
  shared_overhead_rate?: number;
  general_admin_rate?: number;
  target_operating_profit_rate?: number;
  actual_site_personnel_cost?: number | null;
  direct_works?: DirectWorkInput[];
  expense_items?: ExpenseItemInput[] | null;
}

export interface QCDSCalcFields {
  direct_cost_budget: number;
  direct_cost_agreed: number;
  direct_cost_settlement: number;
  labor_insurance: number;
  construction_insurance: number;
  stamp_cost: number;
  receipt_cost: number;
  special_insurance: number;
  site_personnel_cost: number;
  fixed_overhead: number;
  site_overhead_total: number;
  construction_cost_total: number;
  construction_dept_overhead: number;
  shared_overhead: number;
  total_cost: number;
  general_admin_cost: number;
  operating_profit: number;
  operating_profit_rate: number;
  target_operating_profit: number;
}

export interface QCDSResponse {
  id: string;
  project_id: string;
  revision: number;
  spare_cost: number | null;
  industrial_waste_cost: number | null;
  labor_insurance_rate: number;
  construction_insurance_rate: number;
  special_insurance_rate: number;
  office_supplies: number;
  communication_cost: number;
  misc_cost: number;
  site_staff_salary_rate: number;
  common_overhead_rate: number | null;
  shared_overhead_rate: number;
  general_admin_rate: number;
  target_operating_profit_rate: number;
  actual_site_personnel_cost: number | null;
  created_at: string;
  updated_at: string;
  direct_works: DirectWorkRead[];
  expense_items: ExpenseItemRead[];
  calc: QCDSCalcFields;
}

File: backend/app/schemas/qcds.py
"""QCDS関連の Pydantic スキーマ。"""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.models.enums import QCDSCategory

ExpenseSection = Literal["B_site", "B_dept", "C"]


class ExpenseItemInput(BaseModel):
    """経費行の入力データ（PUT リクエスト用）。"""

    id: uuid.UUID | None = None
    section: ExpenseSection
    row_no: int = Field(..., ge=1)
    system_key: str | None = None
    item_name: str = Field(..., max_length=200)
    formula_description: str | None = None
    amount_override: float | None = None
    is_custom: bool = False


class ExpenseItemRead(BaseModel):
    """経費行のレスポンス。effective_amount は override または自動計算値。"""

    model_config = {"from_attributes": True}

    id: uuid.UUID
    qcds_id: uuid.UUID
    section: ExpenseSection
    row_no: int
    system_key: str | None
    item_name: str
    formula_description: str | None
    amount_override: float | None
    is_custom: bool


class DirectWorkInput(BaseModel):
    """直接工事費1行の入力データ。"""

    row_no: int = Field(..., ge=1)
    work_type: str | None = None
    vendor_id: uuid.UUID | None = None
    vendor_name_snapshot: str | None = None
    category: QCDSCategory | None = None
    budget_amount: float | None = None
    agreed_amount: float | None = None
    settlement_amount: float | None = None
    agreement_checked: bool = False
    payment_month_4: float | None = None
    payment_month_5: float | None = None
    payment_month_6: float | None = None
    payment_month_7: float | None = None
    payment_month_8: float | None = None
    payment_month_9: float | None = None
    payment_month_10: float | None = None
    payment_month_11: float | None = None
    payment_month_12: float | None = None
    payment_month_1: float | None = None
    payment_month_2: float | None = None
    payment_month_3: float | None = None
    payment_completed: bool = False
    note: str | None = None


class QCDSInput(BaseModel):
    """QCDS保存リクエスト（ヘッダ + 直接工事費30行）。"""

    revision: int = 0
    spare_cost: float | None = None
    industrial_waste_cost: float | None = None
    labor_insurance_rate: float = 0.001973
    construction_insurance_rate: float = 0.002095
    special_insurance_rate: float = 0.000110
    special_insurance_equipment_rate: float = 0.000110
    special_insurance_demolition_rate: float = 0.019053
    office_supplies: float = 2000
    communication_cost: float = 10000
    misc_cost: float = 5000
    site_staff_salary_rate: float = 0.035
    common_overhead_rate: float | None = None
    shared_overhead_rate: float = 0.05
    general_admin_rate: float = 0.035
    target_operating_profit_rate: float = 0.10
    actual_site_personnel_cost: float | None = None
    direct_works: list[DirectWorkInput] = Field(default_factory=list)
    expense_items: list[ExpenseItemInput] | None = None


class DirectWorkRead(DirectWorkInput):
    """直接工事費1行のレスポンス。"""

    id: uuid.UUID
    vendor_name: str | None = None
    source_scan_result_id: uuid.UUID | None = None


class QCDSCalcFields(BaseModel):
    """QCDS派生計算結果。"""

    direct_cost_budget: float
    direct_cost_agreed: float
    direct_cost_settlement: float
    labor_insurance: float
    construction_insurance: float
    stamp_cost: float
    receipt_cost: float
    special_insurance: float
    special_insurance_equipment: float = 0.0
    special_insurance_demolition: float = 0.0
    site_personnel_cost: float
    fixed_overhead: float
    site_overhead_total: float
    construction_cost_total: float
    construction_dept_overhead: float
    shared_overhead: float
    total_cost: float
    general_admin_cost: float
    operating_profit: float
    operating_profit_rate: float
    target_operating_profit: float


class QCDSResponse(BaseModel):
    """QCDS取得レスポンス（ヘッダ + 直接工事費 + 計算結果）。"""

    id: uuid.UUID
    project_id: uuid.UUID
    revision: int
    spare_cost: float | None
    industrial_waste_cost: float | None
    labor_insurance_rate: float
    construction_insurance_rate: float
    special_insurance_rate: float
    special_insurance_equipment_rate: float = 0.000110
    special_insurance_demolition_rate: float = 0.019053
    office_supplies: float
    communication_cost: float
    misc_cost: float
    site_staff_salary_rate: float
    common_overhead_rate: float | None
    shared_overhead_rate: float
    general_admin_rate: float
    target_operating_profit_rate: float
    actual_site_personnel_cost: float | None
    created_at: datetime
    updated_at: datetime
    direct_works: list[DirectWorkRead]
    expense_items: list[ExpenseItemRead]
    calc: QCDSCalcFields
